lgc/patch/PatchReadFirstLane.cpp

```diff
@@ -0,0 +1,383 @@
+// the definition of the divergent operand that is earliest in basic block order (call it "the earliest divergent
+// operand") and propagate up to that instruction; if it turns out that that instruction can be assumed to be uniform,
+// then we can just insert the readfirstlane there (or propagate).
+//
+// @param block : The process basic block
+void PatchReadFirstLane::findInsertLocations(BasicBlock *block) {
+  m_insertLocations.clear();
+  // Find insert locations or propagate up among "assumed uniforms" by backward traversing the basic block
+  for (auto assumedUniformDivergentUses : m_canAssumeUniformDivergentUsesMap) {
+    Instruction *candidateInst = assumedUniformDivergentUses.first;
+
+    auto &divergentDefSet = assumedUniformDivergentUses.second;
+    const unsigned divergentDefCount = divergentDefSet.size();
+    if (divergentDefCount == 0) {
+      // No divergent use for propagating up and try the candidateInst as an insert location
+      addLocation(candidateInst);
+      continue;
+    }
+    Instruction *earliestDivergentDef = *divergentDefSet.begin();
+    if (divergentDefCount == 1 && m_canAssumeUniformDivergentUsesMap.count(earliestDivergentDef) == 0) {
+      earliestDivergentDef = nullptr;
+    } else if (divergentDefCount > 1) {
+      // Find the earliest divergent definitions in the basic block order
+      for (auto divergentDef : divergentDefSet) {
+        if (divergentDef->comesBefore(earliestDivergentDef))
+          earliestDivergentDef = divergentDef;
+      }
+      // We can propagate up to the earliest divergent operand under two conditions:
+      // 1. The earliest divergent operand definition has only one divergent use
+      // 2. All divergent uses are "assumed uniform"
+      // 3. All divergent uses depend the earliest divergent operand
+      auto earliestUniformDivergentUsesIt = m_canAssumeUniformDivergentUsesMap.find(earliestDivergentDef);
+      bool canPropagateUp = (earliestUniformDivergentUsesIt != m_canAssumeUniformDivergentUsesMap.end() &&
+                             earliestUniformDivergentUsesIt->second.size() == 1);
+      if (canPropagateUp) {
+        DenseSet<Instruction *> divergentUses;
+        for (auto inst : divergentDefSet) {
+          auto assumeUniformOperandsIt = m_canAssumeUniformDivergentUsesMap.find(inst);
+          if (assumeUniformOperandsIt == m_canAssumeUniformDivergentUsesMap.end()) {
+            canPropagateUp = false;
+            break;
+          }
+          for (auto operand : assumeUniformOperandsIt->second) {
+            divergentUses.insert(operand);
+          }
+        }
+        if (canPropagateUp) {
+          Instruction *earliestOperand = *(earliestUniformDivergentUsesIt->second.begin());
+          divergentDefSet.insert(earliestOperand);
+          for (auto use : divergentUses) {
+            if (divergentDefSet.count(use) == 0) {
+              canPropagateUp = false;
+              break;
+            }
+          }
+        }
```

Comment:
Doesn't this have to transitively check instructions somehow? Like in the example @ruiling gave in https://github.com/GPUOpen-Drivers/llpc/pull/976#issuecomment-699591447